### ISSUE TITLE
Vim terminal buffer converts '\0' to new line

### ISSUE THREAD

**Issue by antonioastorino** [vim/vim#16326]
### Steps to reproduce

1. Compile
```c
#include <stdio.h>

int main() {
    for (int i = 0; i < 10; i++) printf("%c", 0);
    printf("some text\n");
    return 0;
}
```
2. Open Vim
3. Launch terminal with `:term`
4. Run compiled binary
All `\0`'s are printed as new lines.

### Expected behaviour

`\0` should not be printed.

### Version of Vim

9.1.950

### Environment

macOS Sonoma, xterm-256color

**Comment by chrisbra** [comment]
That is probably be caused by the restriction mentioned at `:h NL-used-for-Nul`

**Comment by zeertzjq** [comment]
This may be a separate issue from that, as the NUL bytes are coverted to NL much earlier here:
```c
/*
 * Returns the whole buffer contents concatenated for "channel"/"part".
 * Replaces NUL bytes with NL.
 */
    static char_u *
channel_get_all(channel_T *channel, ch_part_T part, int *outlen)
```

**Comment by chrisbra** [comment]
That is exactly the point here. The command output includes the `NUL` bytes and are converted to a new line. Otherwise Vim couldn't handle the NUL bytes properly.

**Comment by zeertzjq** [comment]
> That is exactly the point here. The command output includes the `NUL` bytes and are converted to a new line. Otherwise Vim couldn't handle the NUL bytes properly.

I don't think that's the case. These NUL bytes are not stored in the buffer. They are passed to libvterm via `vterm_input_write()`.

**Comment by chrisbra** [comment]
Well, we can remove it, but not sure this is much better:
```diff
diff --git a/src/channel.c b/src/channel.c
index 69bbff24e..95a796dde 100644
--- a/src/channel.c
+++ b/src/channel.c
@@ -1863,8 +1863,10 @@ channel_get_all(channel_T *channel, ch_part_T part, int *outlen)
     p = res;
     while (p < res + len)
     {
+#if 0
        if (*p == NUL)
            *p = NL;
+#endif
 #ifdef MSWIN
        else if (*p == 0x1b)
        {
```

`find ~/vim-src/src/ -print0` kinda works, but it seems libvterm doesn't print/show all file names. Currently, `find -print0` prints all files by adding new lines, but the cursor position is not reset to the beginning of the line, which is also strange. 

So it looks broken either way and I am not sure a terminal can be expected to print binary NUL values correctly.

**Comment by antonioastorino** [comment]
Can NUL be replaced with something else or skipped?

The issue I faced was twofold.  1) The beginning of the string (the interesting part) was scrolled up too far due to the large number of trailing NULs and 2) I didn't see the same output printed by Terminal.

Because `\0` is the first character in the ASCII, I assume a text-based application should be able to interpret it and print nothing at all. If I remember correctly, I once saw a Windows application (FileMaker) printing them as empty spaces in a text field, but keeping the correct ASCII values, i.e., you could copy those spaces and paste them into any text editor. However, the text editor would print nothing.

I don't know the story behind replacing NUL with NL. if skipping it is not possible, I would suggest replacing it with SP, instead, just to avoid the scrolling issue.

**Comment by chrisbra** [comment]
> Because \0 is the first character in the ASCII, I assume a text-based application should be able to interpret it and print nothing at all. 

Well, that is not true. POSIX defines [text files](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/V1_chap03.html#tag_03_387) as:

> A file that contains characters organized into zero or more lines. The lines do not contain NUL characters and none can exceed {LINE_MAX} bytes in length, including the <newline> character. Although POSIX.1-2024 does not distinguish between text files and binary files (see the ISO C standard), many utilities only produce predictable or meaningful output when operating on text files. The standard utilities that have such restrictions always specify "text files" in their STDIN or INPUT FILES sections.

Most Unix utilities can only be expected to work reliably with text files, they may not work correctly with binary files as expected. I don't know off-hand, if terminals are supposed to handle binary NULs correctly in the output. However, I have often seen terminals to get confused if you accidentally cat a binary archive and you had to use `reset` to reset it to a sane state. 

Binary NULs (and `/` as directory separator for obvious reasons) are guaranteed to be not part of a filename (because in C the binary NUL is the delimiter of character strings, and Unix was based on C).  That's the reason why `find` and `xargs` introduced the `-print0/-0` argument, in order to robustly handle file names. 

Since the NUL is the delimiter for text-strings in C, Vim (as being a C based program), cannot easily handle NUL bytes in memory, therefore they are replaced by line-breaks and that's what the above help file describes. 

Simply replacing the `NUL` bytes by something else is probably not being a good solution, since you are changing the output. I don't like this and it will breaks parsing of the output.

**Comment by zeertzjq** [comment]
> but it seems libvterm doesn't print/show all file names.

That's because `STRLEN()` is used in `write_to_term()`, so the text is truncated at the `NUL`. It may work if the length of the full job output is passed through `channel_get_all() -> may_invoke_callback() -> write_to_term() -> term_write_job_output() -> vterm_input_write()`.

**Comment by ychin** [comment]
>Most Unix utilities can only be expected to work reliably with text files, they may not work correctly with binary files as expected. I don't know off-hand, if terminals are supposed to handle binary NULs correctly in the output. However, I have often seen terminals to get confused if you accidentally cat a binary archive and you had to use reset to reset it to a sane state.

Terminals are generally designed to be resilient to output though. NUL's being used as a terminator in a C is in implementation detail of a C-based program, not necessarily something it should expose to the world. For example, even though Vim is written in C, the whole point of converting to NL is that to an external user, it *looks* like Vim handled the NUL char correctly. The fact that it is converted to NL internally is usually something you don't have to worry about. You will correctly see `^@` in the editor, and if you do `ga` or `g8` on it it will correctly tell you the character is has value 0.

Another case in point is that if your copy-and-paste something in to Vim from the system clipboard, *and* the clipboard has a NUL character, Vim has explicit logic to handle it to convert to NL, which as I already mentioned shows up to the user as a NUL byte. The NL conversion is an internal non-user-facing implementation detail. (this was something my recent PR #16596 had to handle as there's an explicit test case that tests using NUL bytes in registers)

I definitely feel like there's a discrepancy here with how the terminal works. We are exposing this detail to the user, and showing a newline definitely feels like a wrong behavior to me. If you try this in other terminals they all just silently ignore it, which while not great is at least a stable behavior. Also note that terminals are not just for handling text files. The fact that Vim is shipping a terminal means we can't just use POSIX's definition of a text file as reference.

**Comment by chrisbra** [comment]
Yeah, we may need to try to fix this using @zeertzjq [comment](https://github.com/vim/vim/issues/16326#issuecomment-2566363933)
